File: screen-orientation/ios/Sources/ScreenOrientationPlugin/ScreenOrientationPlugin.swift
import Foundation
import UIKit
import Capacitor

@objc(ScreenOrientationPlugin)
public class ScreenOrientationPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ScreenOrientationPlugin"
    public let jsName = "ScreenOrientation"
    public let pluginMethods: [CAPPluginMethod] = [
        .async("orientation", ScreenOrientationPlugin.orientation),
        .promise("lock", ScreenOrientationPlugin.lock),
        .promise("unlock", ScreenOrientationPlugin.unlock)
    ]

    private let implementation = ScreenOrientation()

    override public func load() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.orientationDidChange),
            name: UIDevice.orientationDidChangeNotification,
            object: nil)
        if let viewController = (self.bridge?.viewController as? CAPBridgeViewController) {
            implementation.setCapacitorViewController(viewController)
        }
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    /// UIDevice is UIKit state: the method runs on the main actor, not the bridge queue.
    @MainActor
    public func orientation(_ call: CAPPluginCall) async -> JSObject {
        return ["type": implementation.getCurrentOrientationType()]
    }

    // lock and unlock stay synchronous: the bridge queue runs them in the order of the calls and the implementation
    // hands the UIKit work to the main queue in that order, so a lock after an unlock wins. Async methods would not
    // keep that order.

    public func lock(_ call: CAPPluginCall) throws {
        guard let lockToOrientation = call.getString("orientation") else {
            throw CAPPluginError("Input option 'orientation' must be provided.")
        }
        implementation.lock(lockToOrientation) { error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    public func unlock(_ call: CAPPluginCall) {
        implementation.unlock { error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc private func orientationDidChange() {
        // Ignore changes in orientation if unknown, face up, or face down
        if UIDevice.current.orientation.isValidInterfaceOrientation {
            let orientation = implementation.getCurrentOrientationType()
            notifyListeners("screenOrientationChange", data: ["type": orientation])
        }
    }
}

File: screen-orientation/android/src/main/java/com/capacitorjs/plugins/screenorientation/ScreenOrientationPlugin.kt
package com.capacitorjs.plugins.screenorientation

import android.content.res.Configuration
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "ScreenOrientation")
public class ScreenOrientationPlugin : Plugin() {
    private lateinit var implementation: ScreenOrientation

    override fun load() {
        implementation = ScreenOrientation(activity)
    }

    @PluginMethod
    public fun orientation(call: PluginCall) {
        val ret = JSObject()
        ret.put("type", implementation.currentOrientationType)
        call.resolve(ret)
    }

    @PluginMethod
    public fun lock(call: PluginCall) {
        val orientationType = call.getString("orientation")
        if (orientationType == null) {
            call.reject("Input option 'orientation' must be provided.")
            return
        }
        implementation.lock(orientationType)
        call.resolve()
    }

    @PluginMethod
    public fun unlock(call: PluginCall) {
        implementation.unlock()
        call.resolve()
    }

    override fun handleOnConfigurationChanged(newConfig: Configuration?) {
        super.handleOnConfigurationChanged(newConfig)
        if (newConfig != null && implementation.hasOrientationChanged(newConfig.orientation)) {
            onOrientationChanged()
        }
    }

    private fun onOrientationChanged() {
        val ret = JSObject()
        ret.put("type", implementation.currentOrientationType)
        notifyListeners("screenOrientationChange", ret)
    }
}
